### PR TITLE
[layout] Add hb_ot_layout_get_font_extents()

### PR DESCRIPTION
**Pull request by behdad**
Untested.

Fixes https://github.com/harfbuzz/harfbuzz/issues/3290

### CODE REVIEW

**Comment by behdad** [comment]
Checking whether the NotoCJK fonts have the BASE table values for these...

**Comment by behdad** [comment]
The BASE table also has versions of min/max with a feature-tag, which it says must match the GSUB/GPOS feature tag. That feels non-sense to me. How is one supposed to use it with multiple features?

**Comment by behdad** [comment]
One more thing. Matthias and I discussed adding a "2" version of this, and baseline, API that take `hb_script_t` / `hb_language_t` instead of the OT tags.

**Comment by behdad** [comment]
> Checking whether the NotoCJK fonts have the BASE table values for these...

Nope.

@blueshade7 do you know of any fonts that have BASE table MinMax records?

**Comment by behdad** [comment]
> The BASE table also has versions of min/max with a feature-tag, which it says must match the GSUB/GPOS feature tag. That feels non-sense to me. How is one supposed to use it with multiple features?

Should I expose the feature-tag as well? @khaledhosny

**Comment by blueshade7** [comment]
> @blueshade7 do you know of any fonts that have BASE table MinMax records?

not aware of

**Comment by behdad** [comment]
I did some testing and fixups on this. Seems to work now.

**Comment by behdad** [comment]
Should probably add a test.

**Comment by behdad** [comment]
> Should probably add a test.

Done.

**Comment by behdad** [comment]
> > The BASE table also has versions of min/max with a feature-tag, which it says must match the GSUB/GPOS feature tag. That feels non-sense to me. How is one supposed to use it with multiple features?
> 
> Should I expose the feature-tag as well? @khaledhosny

Just this outstanding.

**Comment by khaledhosny** [comment]
Just a thought, since a line/paragraph can have multiple scripts/language, would it make since for these functions to take an array of scripts/languages and return the maximum extent?

I’m just think to think how can this API be used in multiscript/multilingual text layout. Right now, when multiple fonts are involved, some application will use the maximum of the line metrics of all fonts involved, others I think designate a font as the main font and use its metrics. I probably need how 

This would also be used for feature tags as well (i.e. accepting array of feature tags, not that I can imagine a use case for this, well may be if a stylistic set swishes between accent styles and one of them gives accents that are compressed into the capitals as seen is some signage).

cc @tiroj who have always been interested in this functionality.

**Comment by behdad** [comment]
I think multiple scripts/languages would be higher-level's responsibility.

**Comment by khaledhosny** [comment]
> I think multiple scripts/languages would be higher-level's responsibility.

 Makes sense.

File: src/hb-ot-layout.cc [hunk 1]
```diff
@@ -2036,6 +2036,109 @@ hb_ot_layout_substitute_lookup (OT::hb_ot_apply_context_t *c,
...
+ *
+ * Fetches script/language-specific font extents.  These values are
+ * looked up in the `BASE` table's `MinMax` records.
+ *
+ * If no such extents are found, the default extents for the font are
+ * fetched. As such, the return value of this function can for the
+ * most part be ignored.  Note that the per-script/language extents
+ * do not have a line-gap value, and the line-gap is set to zero in
+ * that case.
+ *
```
Comment (khaledhosny) [inline src/hb-ot-layout.cc:2121]: Might be worth adding a note about the difference between the 2 and the other API.

Comment (behdad) [reply src/hb-ot-layout.cc:2121]: Done.

**Review by khaledhosny — APPROVED** [review]

**Pull request merged**